type-generator/src/main/java/io/mesosphere/types/dtr/compiler/impl/StringTemplateEngine.java

```diff
@@ -0,0 +1,777 @@
+     */
+    private String templateRender(String templateName, BlockArguments arguments) throws IOException {
+        ST tpl = group.getInstanceOf(templateName);
+        if (tpl == null) {
+            throw new IOException("Unable to locate template " + templateName);
+        }
+
+        // Import public getter methods as fields from BlockArguments
+        for (Method method : arguments.getClass().getDeclaredMethods()) {
+            try {
+                if (method.getName().startsWith("get")) {
+                    String fieldName = method.getName().substring(3);
+                    fieldName = fieldName.substring(0, 1).toLowerCase() + fieldName.substring(1);
+                    tpl.add(fieldName, method.invoke(arguments));
+                }
+            } catch (IllegalAccessException e) {
+                e.printStackTrace();
+            } catch (InvocationTargetException e) {
+                e.printStackTrace();
+            }
+        }
+
+        // Import public properties as fields from BlockArguments
+        for (Field field : arguments.getClass().getFields()) {
+            try {
+                tpl.add(field.getName(), field.get(arguments));
+            } catch (IllegalAccessException e) {
+                e.printStackTrace();
+            }
+        }
+
+        // Write template to the output stream
+        return tpl.render();
+    }
+
+//    /**
+//     * Use the project configuration to get the names for the internal types
+//     * (For example internal "string" might be "char *" in C++)
+//     *
+//     * @param type The internal type for which to get the type name
+//     * @return The name of the type as a plain string
+//     */
+//    @Override
+//    public String getTypeName(Type type) {
+//        // Get mapping of scalar types
+//        if (type instanceof ScalarType) {
+//            Value storageType = ((ScalarType) type).storageType;
+//            String storageName = storageType.toName();
+//
+//            // Lookup that value from the language-specific scalar table
+//            return scalarTypes.getOrDefault(storageName, storageName);
+//        }
+//
+//        // Otherwise use the type identifier
+//        return type.getId();
+//    }
+
+    /**
+     * Return the user-defined storage type for the specified value
+     *
+     * @param storageType The internal storage type for which to get a name
+     * @return Returns an expression with the type of the given storage type on the language
+     */
+    @Override
+    public String getStorageTypeName(Value storageType) {
+        String storageName = storageType.toName();
+        return scalarTypes.getOrDefault(storageName, storageName);
+    }
+
+    /**
+     * Render a value template
+     *
+     * @param template
+     * @param type
+     * @param kind
+     * @param value
+     * @return
+     */
+    private String renderValueTemplate(String template, Type type, TypeScope scope, String kind, Object value) {
```

Comment:
What is a value template?

Reply:
It's a template that constructs the language-specific expression that carries the given value. Useful for creating `defaultValue` expressions.

Namely, we have two kinds of templates, that being `inst` or `valueof` and they both accept a very similar set of arguments.